syntax = "proto3";

package common;
option go_package = "github.com/JuanGQCadavid/ds-practice-2025/utils/pb/common";

message User {
    string name = 1;
    string contact = 2;
}

message CreditCard {
    string number = 1;
    string expirationDate = 2;
    string cvv = 3;
}

message Item {
    string name = 1;
    int32 quantity = 2;
}

message Address {
    string street = 1;
    string city = 2;
    string state = 3;
    string zip = 4;
    string country = 5;
}

message Device {
    string type = 1;
    string model = 2;
    string os = 3;
}

message Browser {
    string name = 1;
    string version = 2;
}

message Order {
    User user = 1;
    CreditCard creditCard = 2;
    string userComment = 3;
    repeated Item items = 4;
    string discountCode = 5;
    string shippingMethod = 6;
    string clientCard = 7;
    string giftMessage = 8;
    Address billingAddress = 9;
    bool giftWrapping = 10;
    bool termsAccepted = 11;
    repeated string notificationPreferences = 12;
    Device device = 13;
    Browser browser = 14;
    string appVersion = 15;
    string screenResolution = 16;
    string referrer = 17;
    string deviceLanguage = 18;
}

message InitRequest {
    string orderId = 1;
    Order order = 2;
}

message InitResponse {
    string errMessage = 1;
    bool isValid = 2;
}

message NextRequest {
    string orderId = 1;
    repeated int32 incomingVectorClock = 2;
}

message NextResponse {
    repeated int32 vectorClock = 1;
    string errMessage = 2;
    bool isValid = 3;
}
